source/server/listener_impl.cc

```diff
@@ -350,22 +362,29 @@ ListenerImpl::ListenerImpl(const envoy::config::listener::v3::Listener& config,
       quic_stat_names_(parent_.quicStatNames()),
       missing_listener_config_stats_({ALL_MISSING_LISTENER_CONFIG_STATS(
           POOL_COUNTER(listener_factory_context_->listenerScope()))}) {
-  // All the addresses should be same socket type, so get the first address's socket type is enough.
-  auto address = Network::Address::resolveProtoAddress(config.address());
-  checkIpv4CompatAddress(address, config.address());
-  addresses_.emplace_back(address);
-
-  for (auto i = 0; i < config.additional_addresses_size(); i++) {
-    if (socket_type_ !=
-        Network::Utility::protobufAddressSocketType(config.additional_addresses(i).address())) {
-      throw EnvoyException(fmt::format("listener {}: has different socket type. The listener only "
-                                       "support same socket type for all the addresses.",
-                                       name_));
+  if (config.has_internal_listener()) {
+    addresses_.emplace_back(
+        std::make_shared<Network::Address::EnvoyInternalInstance>(config.name()));
+  } else {
+    // All the addresses should be same socket type, so get the first address's socket type is
+    // enough.
+    auto address = Network::Address::resolveProtoAddress(config.address());
```

Comment:
there should be a case, `config.has_internal_listener()` is false, but internal address is specified in the address field.

Reply:
Oh you mean we don't check for that early enough? We should throw in buildInternalListener but that might be late.

Reply:
yea, I saw you add the check here https://github.com/envoyproxy/envoy/pull/22274/files#diff-b0d5e1a1317df8e22a556ee74d6f728379c5f85ce640a9404acc14bfe8192ad3R569, that is what I said.